OpenSSL/crypto.py

```diff
@@ -1356,6 +1359,125 @@ def add_cert(self, cert):
+
+    :ivar _store: See the ``store`` `__init__`` parameter.
+
+    :ivar _cert: See the ``certificate`` `__init__`` parameter.
+    """
+
+    def __init__(self, store, certificate):
+        """
+        :param X509Store store: The certificates which will be trusted for the
+            purposes of any verifications.
+
+        :param X509 certificate: The certificate to be verified.
+        """
+        store_ctx = _lib.X509_STORE_CTX_new()
+        self._store_ctx = _ffi.gc(store_ctx, _lib.X509_STORE_CTX_free)
+        self._store = store
+        self._cert = certificate
+        # Make the store context available for use after instantiating this
+        # class by initializing it now. Per testing, subsequent calls to
+        # :py:meth:`_init` have no adverse affect.
+        self._init()
+
+
+    def _init(self):
+        """
+        Set up the store context for a subsequent verification operation.
+        """
+        ret = _lib.X509_STORE_CTX_init(self._store_ctx, self._store._store, self._cert._x509, _ffi.NULL)
+        if ret <= 0:
+            _raise_current_error()
+
+
+    def _cleanup(self):
+        """
+        Internally cleans up the store context.
+
+        The store context can then be reused with a new call to
+        :py:meth:`_init`.
+        """
+        _lib.X509_STORE_CTX_cleanup(self._store_ctx)
+
+
+    def _exception_from_context(self):
+        """
+        Convert an OpenSSL native context error failure into a Python
+        exception.
+
+        When a call to native OpenSSL X509_verify_cert fails, additonal information
+        about the failure can be obtained from the store context.
+        """
+        errors = [
+            _lib.X509_STORE_CTX_get_error(self._store_ctx),
+            _lib.X509_STORE_CTX_get_error_depth(self._store_ctx),
+            _native(_ffi.string(_lib.X509_verify_cert_error_string(
+                        _lib.X509_STORE_CTX_get_error(self._store_ctx)))),
+        ]
+        # A context error should always be associated with a certificate, so we
+        # expect this call to never return :class:`None`.
+        _x509 = _lib.X509_STORE_CTX_get_current_cert(self._store_ctx)
+        _cert = _lib.X509_dup(_x509)
+        pycert = X509.__new__(X509)
+        pycert._x509 = _ffi.gc(_cert, _lib.X509_free)
+        return X509StoreContextError(errors, pycert)
+
+
+    def set_store(self, store):
+        """
+        Set the context's trust store.
+
+        :param X509Store store: The certificates which will be trusted for the
+            purposes of any *future* verifications.
+        """
+        self._store = store
+
+
+    def verify_certificate(self):
+        """
+        Verify a certificate in a context.
+
+        :param store_ctx: The :py:class:`X509StoreContext` to verify.
+        :raises: Error
+        """
+        # Always re-initialize the store context in case
+        # :py:meth:`verify_certificate` is called multiple times.
+        self._init()
```

Comment:
Calling _init() again here throws out any changes to verification parameters made between X509Store() and verify_certificate().

Test case:
```python
import time
import calendar

from OpenSSL.crypto import (load_certificate,
                            FILETYPE_PEM,
                            X509Store,
                            X509StoreContext,
                            X509StoreContextError)
from OpenSSL._util import (ffi as _ffi,
                           lib as _lib)
from OpenSSL.test.test_crypto import (root_cert_pem,
                                      intermediate_cert_pem,
                                      intermediate_server_cert_pem)

root_cert = load_certificate(FILETYPE_PEM, root_cert_pem)
intermediate_cert = load_certificate(FILETYPE_PEM, intermediate_cert_pem)
server_cert = load_certificate(FILETYPE_PEM, intermediate_server_cert_pem)

# Root CA with valid time
store = X509Store()
store.add_cert(root_cert)
store_ctx = X509StoreContext(store, root_cert)
store_ctx.verify_certificate()

# Root CA with invalid time
store = X509Store()
store.add_cert(root_cert)
store_ctx = X509StoreContext(store, root_cert)

expire_time = root_cert.get_notAfter().decode('utf-8')
cert_time = time.strptime(expire_time, '%Y%m%d%H%M%SZ')
invalid_time = calendar.timegm(cert_time)
vfy_param = _lib.X509_STORE_CTX_get0_param(store_ctx._store_ctx)
_lib.X509_VERIFY_PARAM_set_time(vfy_param, invalid_time)
try:
    store_ctx.verify_certificate()
except X509StoreContextError as e:
    assert e.args[0][0] == _lib.X509_V_ERR_CERT_HAS_EXPIRED
    assert e.args[0][1] == 0 
```

I thought about modifying verify_certificate() as follows but it is still not perfect because every time you call verify_certificate() it throws away the parameters at the end of the call but at least it won't crash Python 3.4.2 in a segfault with multiple calls.
```python
    def verify_certificate(self):
        """
        Verify a certificate in a context.

        :param store_ctx: The :py:class:`X509StoreContext` to verify.
        :raises: Error
        """
        try:
            ret = _lib.X509_verify_cert(self._store_ctx)
            if ret <= 0:
                raise self._exception_from_context()
        finally:
            self._cleanup()
            # Always re-initialize the store context in case
            # :py:meth:`verify_certificate` is called multiple times.
            self._init()
```